TU20Bot/Commands/NameMatch.cs

```diff
@@ -0,0 +1,154 @@
+﻿using System;
+using System.Threading.Tasks;
+using Discord;
+using Discord.Commands;
+using Discord.WebSocket;
+
+namespace TU20Bot.Commands {
+    public class NameMatch : ModuleBase<SocketCommandContext> {
+
+        private readonly string[,] origNames = {
+            {"Dev","Narula"},
+            {"Timur","Khayrullin"},
+            {"Zaynah","nolastname"},
+            {"Yifan","Wang"},
+            {"Tahmeed","Naser"},
+            {"Shrena","Sribalan"},
+            {"Rick","(O20)"},
+            {"Pranav","Vyas"},
+            {"Karen","Truong"},
+            {"Muhammad Muizz","nolastname"},
+            {"Vince","Li"},
+            {"Alex","Li"},
+            {"nofirstname","Nolan"},
+            {"Borna","Sadeghi"},
+            {"Hargun","nolastname"},
+            {"Markos","Georghiades"},
+            {"Muhammad Muizz","Zafar"},
+            {"Taylor desgroup","Whatley"},
+            {"Viktor fasest","Korolyuk"},
+            {"Denys","Linkov"},
+            {"Muhammad Ali","Syed"},
+        };
+
+        [Command("lsuser-assignrole")]
+        [Summary("matches guild member names with a specified list and assigns roles to the matched names")]
+        public async Task matchNames(ulong roleId) {
+            var users = Context.Guild.Users;
+            string discordMessage = "";
+            string noLastNameMsg = "";
+            string noFirstNameMsg = "";
+            string notInGuildMsg = "";
+            for (int i = 0; i < origNames.GetLength(0); i++) {
+                foreach (var user in users) {
```

Comment:
But wouldn't this way be a bit slower than what we currently have?

Reply:
Why do you think that? 🙂 

My theory is that every time we get a user object out of the users list, the value is queried from discord. If this is not true I suppose they are of equal performance but this is what experimentation is for!! :D 

Let our lists be
```
users -> user1, user2, user3
origNames -> name1, name2, name3
```

Tracing of the existing loop:
```
Getting name1 from list
Getting user1 from discord
Getting user2 from discord
Getting user3 from discord
Getting name2 from list
Getting user1 from discord
Getting user2 from discord
Getting user3 from discord
Getting name3 from list
...
```

Tracing of the suggested loop (theoretically):
```
Getting user1 from discord
Getting name1 from list
Getting name2 from list
Getting name3 from list
Getting user2 from discord
Getting name1 from list
Getting name2 from list
Getting name3 from list
Getting user3 from discord
...
```

I assume in the second on there are less requests to discord -> more performance + rate limit runs out slower?

Reply:
I was looking at it from more of a total number of times the loop iterates. Since the list of users, we get from discord is bigger than the list of users we externally provide the bot with. 
```
users -> user1, user2, user3, user4, user5, user6, user7
origNames -> name1, name2
```

```
Getting name1 from list
Getting user1 from discord (name doesn't match)
Getting user2 from discord (name doesn't match)
Getting user3 from discord (name doesn't match)
Getting user4 from discord (name doesn't match)
Getting user5 from discord (name matches)

break the loop and get the next user in the list
```

```
Getting user1 from discord
Getting name1 from the list(name doesn't match)
Getting name2 from the list(name doesn't match)

Getting user2 from discord
Getting name1 from the list(name doesn't match)
Getting name2 from the list(name doesn't match)

Getting user3 from discord
Getting name1 from the list(name doesn't match)
Getting name2 from the list(name doesn't match)

Getting user4 from discord
Getting name1 from the list(name doesn't match)
Getting name2 from the list(name doesn't match)

Getting user5 from discord
Getting name1 from the list(name matches)

break the loop and get the next user from discord
```

Reply:
But I'll try both of the methods and see which one is better. Also, I didn't think I would be querying discord every time I get a user in the for loop bcz I thought that I already got the users list from discord and that was the end of the "communicating with discord" part. Do you remember the message .net was giving me: `A MessageReceived handler is blocking the gateway task.` This might be the rzn why.

Reply:
> This might be the rzn why.
Yup that's why I brought this up :D

> Since the list of users, we get from discord is bigger than the list of users we externally provide the bot with.
A couple cool theory things!~

 - Since the order of the lists are undefined, both your solution and my solution perform identically. It's a matter of chance if the first user breaks on the first iteration or last iteration (depending on how the list is ordered). Neither will outperform the other (if both lists have O(1) fetch time).
 - You can define C# properties like this that will only get the information you need. I think it's possible that every iteration of that loop it will make a new request, although I am not sure.